### ISSUE TITLE
GraphQL queries run in previews don't seem to be honouring GraphQL aliases or schema customization

### ISSUE THREAD

**Issue by tjjjwxzq** [angeloashmore/gatsby-source-prismic#378]
I've been having quite a bit of trouble getting previews to work smoothly, and I notice they keep breaking because certain data fields, that are queried perfectly fine when the component is loaded normally (ie. not through preview), seem to become undefined and blow-up the preview.

For eg.: I see this in preview mode, but it renders fine normally:
<img width="1096" alt="image" src="https://user-images.githubusercontent.com/11689299/119764551-34a54780-bee4-11eb-9d0a-c7f246e7e536.png">

For the error about `buttons` being undefined, I've managed to triangulate it to the fact that data that is being passed into the component during preview does not seem to name a field according to the GraphQL alias, but retains the original field name. That is, there is a field called `items` that I alias to `buttons` in my query, but this does not seem to work in preview:

Object logged out in preview:
<img width="431" alt="image" src="https://user-images.githubusercontent.com/11689299/119764679-70401180-bee4-11eb-8684-1fae30fe43ee.png">

Relevant portion of my GraphQL query:
<img width="439" alt="image" src="https://user-images.githubusercontent.com/11689299/119764717-7e8e2d80-bee4-11eb-8b64-de614969546e.png">

On a different page, where I'm using the `createResolvers` API to add a new `contributors` field on one of my types:
<img width="678" alt="image" src="https://user-images.githubusercontent.com/11689299/119764799-a7aebe00-bee4-11eb-9fe2-63fc86675b1a.png">

The `contributors` field doesn't seem to be populated correctly in preview mode, but I can query it fine normally and in the GraphQL playground.

Broken in preview:
<img width="451" alt="image" src="https://user-images.githubusercontent.com/11689299/119764882-cad96d80-bee4-11eb-983a-f771d934f739.png">

But totally fine when querying in the playground:
<img width="913" alt="image" src="https://user-images.githubusercontent.com/11689299/119764915-d7f65c80-bee4-11eb-82a9-5ecf8a90302c.png">

I suspect this is a bug which how this plugin is handling the merging of preview data with component data, but I don't know enough about the internals of what the preview HOC components are doing to know for certain.

Don't mind spending some time investigating and maybe opening a PR to fix, but wondering if the maintainers (@angeloashmore ?) with more context could help to point me in the right direction? And whether there have been similar issues reported before (I tried to dig through historical issues, but couldn't find anything exactly like this about broken previews).

**Comment by jodiedoubleday** [comment]
I'm trying to find the issue where this was mentioned but previews don't support aliases.
I believe this is because this plugin uses the RestAPI and uses that for previews so has no knowledge of the aliases you have made.
This may be fixed in v4 potentially.

**Comment by tjjjwxzq** [comment]
Thanks @jodiedoubleday for the response. I guess aliases aren't too big an issue for me since they can be circumvented easily with a separate mapping function to transform the query data.

But the null data with the customized schema worries me, there isn't an elegant workaround for that other than bringing the join operation into runtime code, which I'd rather avoid.

**Comment by angeloashmore** [comment]
Thanks @jodiedoubleday, that's correct. Field aliasing won't work during previews since the GraphQL server Gatsby uses during development/build time is no longer there.

This document explains some of the preview limitations. It is written for the upcoming V4 release, but still applies to V3.
https://github.com/angeloashmore/gatsby-source-prismic/blob/beta/packages/gatsby-plugin-prismic-previews/docs/limitations.md

Your schema customizations fail for the same reason. Modifying the GraphQL schema in Gatsby only affects the development and build time API. During previews, we are only working in the browser without Gatsby.

However, what you are setting up could probably be accomplished without creating a schema customization. If your document has a `contributors` field that links to Contributor documents, the GraphQL API already returns those documents. This assumes your `contributors` field is a Group field containing a `contributor` Content Relation field.

See this part of the documentation: https://github.com/angeloashmore/gatsby-source-prismic#query-content-relation-fields

Your query could look something like the following:

```graphql
query MyQuery {
  allPrismicArticle {
    nodes {
      id
      data {
        contributors {
          contributor {
            document {
              ... on PrismicContributor {
                id
              }
            }
          }
        }
      }
    }
  }
}
```

**Comment by tjjjwxzq** [comment]
Thanks @angeloashmore for the detailed response and suggested workarounds! Still getting to grips with GraphQL so didn't realize the foreign key join could be approached that way. Will give it a shot and update and close this issue if all goes well, for any others who struggle with the same problems.

**Comment by tjjjwxzq** [comment]
Alright, I've got previews to work by removing GraphQL aliases through:

1. Transforming the data via a separate data
2. Splitting a query for which the above wouldn't work (I needed to query the same field with two different parameters) out into separate `useStaticQuery` calls in two separte hooks files which I then imported and used in my component

And following the recommended approach of querying documents linked by a content relationship field worked like a charm. 

I wonder if the limitations doc should be amended slightly to explicitly mention that schema customization is not supported as well? (Although that conclusion follows naturally from the fact that things are happening at runtime not build time, that's not necessarily obvious to everyone)
